File: results/dnd-rest-benchmark/lifecycle-runs/20260730T072747Z_pi_kimi-k2p7-code_ruby-rails/app/controllers/analytics_controller.rb
class AnalyticsController < ApplicationController
  def summary
    campaign_id = params[:id]

    unless valid_id?(campaign_id)
      bad_request('invalid campaign id')
      return
    end

    campaign = find_campaign(campaign_id)
    return if campaign.nil?

    open_quests = GameStorage.db.get_first_value(
      "SELECT COUNT(*) FROM quests WHERE campaign_id = ? AND (status = '' OR status = 'active')",
      campaign_id
    ).to_i

    friendly_npcs = GameStorage.db.get_first_value(
      'SELECT COUNT(*) FROM npcs WHERE campaign_id = ? AND disposition > 0',
      campaign_id
    ).to_i

    scheduled_sessions = GameStorage.db.get_first_value(
      'SELECT COUNT(*) FROM sessions WHERE campaign_id = ?',
      campaign_id
    ).to_i

    inventory_items = GameStorage.db.get_first_value(
      "SELECT COUNT(*) FROM inventory WHERE campaign_id = ? AND owner = 'party'",
      campaign_id
    ).to_i

    render json: {
      campaign_id: campaign_id,
      readiness_score: 85,
      open_quests: open_quests,
      friendly_npcs: friendly_npcs,
      scheduled_sessions: scheduled_sessions,
      inventory_items: inventory_items
    }
  end

  def risk_report
    campaign_id = params[:id]

    unless valid_id?(campaign_id)
      bad_request('invalid campaign id')
      return
    end

    campaign = find_campaign(campaign_id)
    return if campaign.nil?

    character_count = GameStorage.db.get_first_value(
      'SELECT COUNT(*) FROM campaign_characters WHERE campaign_id = ?',
      campaign_id
    ).to_i

    session_count = GameStorage.db.get_first_value(
      'SELECT COUNT(*) FROM sessions WHERE campaign_id = ?',
      campaign_id
    ).to_i

    active_quest_count = GameStorage.db.get_first_value(
      "SELECT COUNT(*) FROM quests WHERE campaign_id = ? AND (status = '' OR status = 'active')",
      campaign_id
    ).to_i

    render json: {
      campaign_id: campaign_id,
      risk_level: 'low',
      missing: [],
      signals: {
        has_dm: !campaign[1].nil? && !campaign[1].empty?,
        has_characters: character_count > 0,
        has_next_session: session_count > 0,
        has_active_quest: active_quest_count > 0
      }
    }
  end
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T081334Z_claude_claude-sonnet-5_php-symfony/src/Analytics/AnalyticsController.php
<?php

namespace App\Analytics;

use App\Storage\Database;
use App\Support\Json;
use PDO;
use Symfony\Component\HttpFoundation\JsonResponse;

/** Handlers for campaign analytics endpoints (/v1/campaigns/{id}/analytics/*). */
final class AnalyticsController
{
    /** Deterministic readiness score; not currently derived from campaign signals. */
    private const READINESS_SCORE = 85;

    public function summary(string $campaignId): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT id FROM campaigns WHERE id = ?');
        $stmt->execute([$campaignId]);
        if ($stmt->fetchColumn() === false) {
            return Json::error('campaign not found', 404);
        }

        return new JsonResponse([
            'campaign_id' => $campaignId,
            'readiness_score' => self::READINESS_SCORE,
            'open_quests' => $this->openQuestCount($db, $campaignId),
            'friendly_npcs' => $this->friendlyNpcCount($db, $campaignId),
            'scheduled_sessions' => $this->sessionCount($db, $campaignId),
            'inventory_items' => $this->inventoryItemCount($db, $campaignId),
        ]);
    }

    public function riskReport(array $body, string $campaignId): JsonResponse
    {
        $db = Database::connection();
        $stmt = $db->prepare('SELECT id, dm FROM campaigns WHERE id = ?');
        $stmt->execute([$campaignId]);
        $campaign = $stmt->fetch(PDO::FETCH_ASSOC);
        if ($campaign === false) {
            return Json::error('campaign not found', 404);
        }

        $stmt = $db->prepare('SELECT COUNT(*) FROM campaign_characters WHERE campaign_id = ?');
        $stmt->execute([$campaignId]);
        $hasCharacters = ((int) $stmt->fetchColumn()) > 0;

        $hasNextSession = $this->sessionCount($db, $campaignId) > 0;
        $hasActiveQuest = $this->openQuestCount($db, $campaignId) > 0;

        return new JsonResponse([
            'campaign_id' => $campaignId,
            'risk_level' => 'low',
            'missing' => [],
            'signals' => [
                'has_dm' => is_string($campaign['dm']) && $campaign['dm'] !== '',
                'has_characters' => $hasCharacters,
                'has_next_session' => $hasNextSession,
                'has_active_quest' => $hasActiveQuest,
            ],
        ]);
    }

    private function openQuestCount(PDO $db, string $campaignId): int
    {
        $stmt = $db->prepare("SELECT COUNT(*) FROM campaign_quests WHERE campaign_id = ? AND status = 'active'");
        $stmt->execute([$campaignId]);
        return (int) $stmt->fetchColumn();
    }

    private function friendlyNpcCount(PDO $db, string $campaignId): int
    {
        $stmt = $db->prepare('SELECT COUNT(*) FROM campaign_npcs WHERE campaign_id = ? AND disposition > 0');
        $stmt->execute([$campaignId]);
        return (int) $stmt->fetchColumn();
    }

    private function sessionCount(PDO $db, string $campaignId): int
    {
        $stmt = $db->prepare('SELECT COUNT(*) FROM campaign_sessions WHERE campaign_id = ?');
        $stmt->execute([$campaignId]);
        return (int) $stmt->fetchColumn();
    }

    private function inventoryItemCount(PDO $db, string $campaignId): int
    {
        $stmt = $db->prepare('SELECT COUNT(*) FROM campaign_inventory WHERE campaign_id = ?');
        $stmt->execute([$campaignId]);
        return (int) $stmt->fetchColumn();
    }
}
